text/0000-api-evolution.md

```diff
@@ -0,0 +1,748 @@
+
+Note that adding new public items is currently a breaking change, due to glob
+imports. For example, the following snippet of code will break if the `foo`
+module introduces a public item called `bar`:
+
+```rust
+use foo::*;
+fn bar() { ... }
+```
+
+The problem here is that glob imports currently do not allow any of their
+imports to be shadowed by an explicitly-define item.
+
+There are two reasons this is considered a minor change by this RFC:
+
+1. The RFC also suggests permitting shadowing of a glob import by any explicit
+   item. This has been the intended semantics of globs, but has not been
+   implemented. The details are left to a future RFC, however.
+
+2. Even if that change were made, though, there is still the case where two glob
+   imports conflict without any explicit definition "covering" them. This is
+   permitted to produce an error under the principles of this RFC because the
+   glob imports could have been written as more explicit (expanded) `use`
+   statements. It is also plausible to do this expansion automatically for a
+   crate's dependencies, to prevent breakage in the first place.
+
+### Structs
+
+See "[Signatures in type definitions](#signatures-in-type-definitions)" for some
+general remarks about changes to the actual types in a `struct` definition.
+
+#### Major change: adding a private field when all current fields are public.
+
+This change has the effect of making external struct literals impossible to
+write, which can break code irreparably.
+
+#### Major change: adding a public field when no private field exists.
+
+This change retains the ability to use struct literals, but it breaks existing
+uses of such literals; it likewise breaks exhaustive matches against the struct.
+
+#### Minor change: adding or removing private fields when at least one already exists.
+
+No existing code could be relying on struct literals for the struct, nor on
+exhaustively matching its contents, and client code will likewise be oblivious
+to the addition of further private fields.
+
+For tuple structs, this is only a minor change if furthermore *all* fields are
+currently private. (Tuple structs with mixtures of public and private fields are
+bad practice in any case.)
+
+#### Minor change: going from a tuple struct with all private fields (with at least one field) to a normal struct, or vice versa.
+
+This is technically a breaking change:
+
+```rust
+// in some other module:
+pub struct Foo(SomeType);
+
+// in downstream code
+let Foo(_) = foo;
+```
+
+Changing `Foo` to a normal struct can break code that matches on it -- but there
+is never any real reason to match on it in that circumstance, since you cannot
+extract any fields or learn anything of interest about the struct.
+
+### Enums
+
+See "[Signatures in type definitions](#signatures-in-type-definitions)" for some
+general remarks about changes to the actual types in an `enum` definition.
+
+#### Major change: adding new variants.
```

Comment:
`ErrorKind` has an unstable variant that prevents stable code from exhaustively matching on it: https://github.com/rust-lang/rust/blob/master/src/libstd/io/error.rs#L118-L123